src/lib/rule2F_test.ts

```diff
@@ -0,0 +1,215 @@
+    render(
+      html`
+        <div id="foo">
+          Hello
+          <div aria-labelledby="bar"></div>
+        </div>
+        <div id="bar">
+          world
+          <div aria-labelledby="foo"></div>
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    context.isLabelReference = true;
+    expect(rule2F(elem!, context)).toBe('Hello world');
+  });
+});
+
+describe('The function allowsNameFromContent', () => {
+  let container: HTMLElement;
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  it('returns true for roles that allow name from content', () => {
+    render(
+      html`
+        <div id="foo" role="button">
+          Hello world
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('returns false for roles that do not allow name from content', () => {
+    render(
+      html`
+        <div id="foo" role="presentation">
+          Hello world
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(false);
+  });
+
+  it('recognises semantic html elements', () => {
+    render(
+      html`
+        <h1 id="foo">
+          Hello world
+        </h1>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('allows table calls only if they are within a table', () => {
+    render(
+      html`
+        <table>
+          <td id="foo"></td>
+        </table>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('allows options only if they are within a datalist or select', () => {
+    render(html` <option id="foo"></option> `, container);
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(false);
+  });
+
+  it('allows options only if they are within a datalist or select', () => {
+    render(
+      html`
+        <select>
+          <option id="foo"></option>
+        </select>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('allows options only if they are within a datalist or select', () => {
+    render(
+      html`
+        <datalist>
+          <option id="foo"></option>
+        </datalist>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(allowsNameFromContent_TEST(elem!)).toBe(true);
+  });
+
+  it('allows inputs of certain types', () => {
```

Comment:
See comment above, I think `returns true for an input of type "button"` is more precise.